src/codegen/codegen_compatibility_visitor.cpp

```diff
@@ -0,0 +1,142 @@
+                                                       AstNodeType::DISCRETE_BLOCK,
+                                                       AstNodeType::FUNCTION_TABLE_BLOCK,
+                                                       AstNodeType::INDEPENDENT_BLOCK,
+                                                       AstNodeType::GLOBAL_VAR,
+                                                       AstNodeType::POINTER_VAR,
+                                                       AstNodeType::BBCORE_POINTER_VAR};
+    incompatible_ast_nodes = AstLookupVisitor().lookup(node, incompatible_ast_types);
+
+    std::stringstream ss;
+    for (auto it: incompatible_ast_nodes) {
+        if (it->is_solve_block()) {
+            auto solve_block = dynamic_cast<ast::SolveBlock*>(it.get());
+            auto method = solve_block->get_method();
+            if (method != nullptr && method->get_node_name() != "cnexp" &&
+                method->get_node_name() != "euler" && method->get_node_name() != "derivimplicit" &&
+                method->get_node_name() != "sparse") {
+                ss << "\"" << method->get_node_name() << "\" solving method used at [";
+                ss << method->get_token()->position() << "] not supported. ";
+                ss << "Supported methods are cnexp, euler, derivimplicit and sparse\n";
+            }
+        } else if (it->is_discrete_block() || it->is_partial_block()) {
+            std::string node_type_name = it->get_node_type_name();
+            // remove "Block" substring
+            node_type_name.erase(node_type_name.end() - 5, node_type_name.end());
+            // transform name of node type to capital letters, as it is in mod files
+            std::transform(node_type_name.begin(),
+                           node_type_name.end(),
+                           node_type_name.begin(),
+                           ::toupper);
+
+            ss << "\"" << it->get_node_name() << "\" " << node_type_name << " construct found at [";
+            ss << it->get_token()->position() << "] is not supported\n";
+        } else if (it->is_ba_block()) {
+            ss << "BEFORE/AFTER construct found at [" << it->get_token()->position();
+            ss << "] is not supported\n";
+        } else if (it->is_terminal_block() || it->is_match_block() || it->is_constant_block() ||
+                   it->is_constructor_block() || it->is_destructor_block() ||
+                   it->is_independent_block()) {
+            std::string node_type_name = it->get_node_type_name();
+            // remove "Block" substring
+            node_type_name.erase(node_type_name.end() - 5, node_type_name.end());
+            // transform name of node type to capital letters, as it is in mod files
+            std::transform(node_type_name.begin(),
+                           node_type_name.end(),
+                           node_type_name.begin(),
+                           ::toupper);
```

Comment:
In general, if something is repeated, it should either go to utils or local lambda function. But again, this should be replaced by get_nmodl_name().